mlrun/runtimes/base.py

```diff
@@ -941,19 +943,32 @@ def _update_run_state(
 
         updates = None
         last_state = get_in(resp, "status.state", "")
+        kind = get_in(resp, "metadata.labels.kind", "")
         if last_state == "error" or err:
-            updates = {"status.last_update": now_date().isoformat()}
-            updates["status.state"] = "error"
+            updates = {
+                "status.last_update": now_date().isoformat(),
+                "status.state": "error",
+            }
             update_in(resp, "status.state", "error")
             if err:
                 update_in(resp, "status.error", err_to_str(err))
             err = get_in(resp, "status.error")
             if err:
                 updates["status.error"] = err_to_str(err)
+
         elif not was_none and last_state != "completed":
-            updates = {"status.last_update": now_date().isoformat()}
-            updates["status.state"] = "completed"
-            update_in(resp, "status.state", "completed")
+
+            # TODO: add a 'workers' section in run objects state, each worker will update its state while
+            #  the run state will be resolved by the server.
+            if kind != "mpijob":
+                logger.debug(
+                    "Updating run state to completed", kind=kind, last_state=last_state
+                )
+                updates = {
+                    "status.last_update": now_date().isoformat(),
+                    "status.state": "completed",
+                }
```

Comment:
not a fan of the `base` knowing about a special case of a class that inherits from it. I would suggest using some helper function for that ( where most of the classes will have the base implementation and the mpi a different one )

Reply:
I agree with you but we can't do this 
keep in mind when you are running on the mpi worker you don't know your kind is mpi (your actual kind is local)

Reply:
but you check that your kind isn't `mpijob` so I am not sure I follow why you can't resolve that from the kind